### PR TITLE
Add Anthropic ChatModel to langchain

### PR DESCRIPTION
**Pull request by mikelambert**
* Adds an Anthropic ChatModel
* Factors out common code in our LLMModel and ChatModel
* Supports streaming  llm-tokens to the callbacks on a delta basis (until a future V2 API does that for us)
* Some fixes

### CODE REVIEW

**Comment by mikelambert** [comment]
Stuck on these two lint errors:
```
tests/integration_tests/chat_models/test_anthropic.py:59: error: "Generation" has no attribute "message"  [attr-defined]
tests/integration_tests/chat_models/test_anthropic.py:60: error: "Generation" has no attribute "message"  [attr-defined]
```

Looks like `LLMResult` and `ChatResult` don't share a common class, and `BaseChatModel` is typed to return a `LLMResult` (due to the superclass) in a bunch of places but returns `ChatResult`. And `on_llm_end` takes an LLMResult and not a `ChatResult`. Looks like it'd need a larger refactor to fix all these, which I suspect is out of scope of this PR.

File: langchain/chat_models/anthropic.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+class AnthropicChat(BaseChatModel, _AnthropicCommon):
+    r"""Wrapper around Anthropic's large language model.
+
+    To use, you should have the ``anthropic`` python package installed, and the
+    environment variable ``ANTHROPIC_API_KEY`` set with your API key, or pass
+    it as a named parameter to the constructor.
+
+    Example:
+        .. code-block:: python
+            import anthropic
```
Comment (vowelparrot) [inline langchain/chat_models/anthropic.py:26]:
```suggestion
```

File: tests/integration_tests/chat_models/test_anthropic.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        streaming=True,
+        callback_manager=callback_manager,
+        verbose=True,
+    )
+    chat_messages: list[BaseMessage] = [HumanMessage(content="How many toes do dogs have?")]
+    result: LLMResult = await chat.agenerate([chat_messages])
+    assert callback_handler.llm_streams > 1
+    assert isinstance(result, LLMResult)
+    for response in result.generations[0]:
+        assert isinstance(response.message, AIMessage)
```
Comment (vowelparrot) [inline tests/integration_tests/chat_models/test_anthropic.py:59]: The OpenAI tests look like

```
assert isinstance(response, ChatGeneration)
assert isinstance(response.text, str)
assert response.text == response.message.content
```

which should fix the listing complaints.

The point around fixing the chat and LLM result inheritance is a good one and agree it's beyond the scope of this PR

**Review by vowelparrot — APPROVED** [review]

File: langchain/chat_models/anthropic.py (hunk 1 above)
Comment (mikelambert) [reply langchain/chat_models/anthropic.py:26]: This import is necessary for the use of `anthropic.HUMAN_PROMPT` down below.

File: tests/integration_tests/chat_models/test_anthropic.py (hunk 2 above)
Comment (mikelambert) [reply tests/integration_tests/chat_models/test_anthropic.py]: Sounds good. Rewritten as such.

**Comment by mikelambert** [comment]
Various struggles with lint/black/etc, but I believe this is in a good state now for review/merging. Updated the model name and re-added some stop-token stuff as well.

**Review by agola11 — APPROVED** [review]
This looks great, thanks @mikelambert! One small nit is that we currently call our chat model implementation of OpenAI `ChatOpenAI` -- might be good to follow that pattern and call this `ChatAnthropic` over `AnthropicChat`

**Comment by mikelambert** [comment]
Thanks, fixed!

**Pull request merged**